### PR TITLE
IFTAQCAA-4 - test script to cover checkout functionality

### PR DESCRIPTION
**Pull request by PKuravskyi**
Added test to cover Checkout functionality

### CODE REVIEW

File: src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    public CheckoutPage inputCustomerData() {
+        surnameFieldElement.setValue(getCustomerOrderData().getSurname());
+        nameFieldElement.setValue(getCustomerOrderData().getName());
+        phoneFieldElement.setValue(getCustomerOrderData().getPhone());
+        var customerCity = cityFieldElement.setValue(getCustomerOrderData().getCity());
+
+        if (cityDropdownElement
+                .getText()
+                .contains(customerCity.getText())) {
+            cityDropdownElement.click();
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java:42]: take a look onto this xpath //li[contains(@class,'autocomplete__item')][normalize-space()='Львів'], you use it as template and pass here the city

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 2]
```diff
@@ -16,7 +17,53 @@ public void verifyLanguageSwitchingTest() {
...
     }
+
+    @Test
+    public void verifyCheckoutPageFunctionalityTest() {
+        homePage.closeAdBanner();
+        var lastProductOfSection = homePage.getLastProductOfSection("Акційні пропозиції");
+
+        assertThat(lastProductOfSection)
+                .as("Last product of section should be opened")
+                .isEqualTo(productPage.getProductName());
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:31]: what are you checking here? in assert should be verified something from page with some expected value

File: src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    public String getPhoneFieldText() {
+        return phoneFieldElement.getValue();
+    }
+
+    public String getCityFieldText() {
+        return cityFieldElement.getValue();
+    }
+
+    public CheckoutPage inputCustomerData() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java:33]: setCustomerData

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 4]
```diff
@@ -16,7 +17,53 @@ public void verifyLanguageSwitchingTest() {
...
+
+        productPage
+                .addProductToBasket()
+                .goToCheckout();
+
+        checkoutPage.inputCustomerData();
+
+        var expectedSurname = getCustomerOrderData().getSurname();
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:39]: extract getCustomerOrderData() into variable

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 5]
```diff
@@ -16,7 +17,53 @@ public void verifyLanguageSwitchingTest() {
...
+        assertThat(checkoutPage.getCityFieldText())
+                .as("City should be '" + expectedCity + "'")
+                .contains(expectedCity);
+
+        checkoutPage
+                .editOrder()
+                .addOneMoreProduct()
+                .goToCheckout();
+
+        var productAmountPrice = checkoutPage.getProductAmountPrice();
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:64]: maybe totalPrice, amount is not clear, it's a synonym of quantity

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/test/java/com/softserveinc/ita/HomePageTest.java (hunk 2 above)
Comment (PKuravskyi) [reply src/test/java/com/softserveinc/ita/HomePageTest.java:31]: method getLastProductOfSection (variable lastProductOfSection) clicks on last product and returns its name. Im verifying that this name equals to product's name we opened

File: src/main/java/com/softserveinc/ita/pageobjects/BasePage.java [hunk 6]
```diff
@@ -63,9 +62,7 @@ public boolean isLanguageSwitchedTo(LanguageSwitcher language) {
     }
 
     public T closeAdBanner() {
-        if (adBannerCrossIconElement.is(visible)) {
-            adBannerCrossIconElement.click();
-        }
+        $("span .exponea-close-cross").shouldBe(visible).click();
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/BasePage.java:65]: chain it please)

File: src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java [hunk 7]
```diff
@@ -59,18 +62,25 @@ public int getProductPrice() {
...
 
     public int getProductQuantity() {
-        var productQuantity = $(".js-product-quantity dd")
-                .getText();
+        var productQuantity = $(".js-product-quantity dd").getText();
 
         return parseInt(productQuantity);
     }
+
+    public boolean didCheckoutPageOpenCorrectly() {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java:79]: it will always return true, what purpose of this method?

File: src/test/java/com/softserveinc/ita/HomePageTest.java [hunk 8]
```diff
@@ -21,6 +26,9 @@ public void verifyLanguageSwitchingTest() {
                 .isTrue();
     }
 
+    protected ProductPage productPage = new ProductPage();
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/HomePageTest.java:29]: init this pages in every test where they are needed

between tests should not be any variable or method

File: src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java [hunk 9]
```diff
@@ -30,16 +31,18 @@ public String getCityFieldText() {
         return cityFieldElement.getValue();
     }
 
-    public CheckoutPage inputCustomerData() {
-        surnameFieldElement.setValue(getCustomerOrderData().getSurname());
-        nameFieldElement.setValue(getCustomerOrderData().getName());
-        phoneFieldElement.setValue(getCustomerOrderData().getPhone());
-        var customerCity = cityFieldElement.setValue(getCustomerOrderData().getCity());
+    public CheckoutPage setCustomerData() {
+        var customerOrderData = getCustomerOrderData();
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java:35]: my mistake, please pass any test data to pages through methods params

File: src/main/java/com/softserveinc/ita/utils/runners/TestRunner.java [hunk 10]
```diff
@@ -1,23 +1,17 @@
...
 
-    protected HomePage homePage = new HomePage();
-    protected ProductPage productPage = new ProductPage();
-    protected CheckoutPage checkoutPage = new CheckoutPage();
-
     @BeforeMethod
     public void openHomePage() {
         Configuration.browserSize = "1920x1080";
-        Configuration.timeout = 8000;
+        Configuration.timeout = ofSeconds(8).toMillis();
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/utils/runners/TestRunner.java:14]: I mean, you also could add  this one: Configuration.pageLoadTimeout = ofSeconds(300).toMillis();
but first read about this config and maybe share info with others

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/utils/runners/TestRunner.java (hunk 10 above)
Comment (PKuravskyi) [reply src/main/java/com/softserveinc/ita/utils/runners/TestRunner.java:14]: Default value for PageLoadTimeout is 30 seconds, this is enough for me. But default value for timeout (4 sec) is not enough because then i often get "Element not found" error.

File: src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java (hunk 7 above)
Comment (PKuravskyi) [reply src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java:80]: image

File: src/main/java/com/softserveinc/ita/pageobjects/HomePage.java [hunk 11]
```diff
@@ -1,5 +1,18 @@
 package com.softserveinc.ita.pageobjects;
 
+import static com.codeborne.selenide.Selenide.$x;
+import static java.lang.String.format;
+
 public class HomePage extends BasePage<HomePage> {
 
+    public String getLastProductOfSection(String sectionName) {
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/HomePage.java:8]: probably create a separate method for getting the title of the last product of section and method which would open specific product in section by title and return product page

File: src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java (hunk 7 above)
Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java:78]: rename to refreshIfErrorExists()

move to base page

return statement generic

you can try do it through while

var count = 0;

while($("rz-critical-error.ng-star-inserted").exists() && count < 5){
refresh();
}

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+import com.softserveinc.ita.pageobjects.ProductPage;
+import com.softserveinc.ita.utils.runners.TestRunner;
+import org.testng.annotations.Test;
+
+import static com.softserveinc.ita.repos.CustomerOrderDataRepo.getCustomerOrderData;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CheckoutPageTest extends TestRunner {
+
+    protected HomePage homePage = new HomePage();
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/CheckoutPageTest.java:14]: leave only home page variable as class var, make it private final

others should be inited in a test near the places where they being needed or get they through methods return statements

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        var lastProductOfSection = homePage.getLastProductOfSection("Акційні пропозиції");
+
+        assertThat(lastProductOfSection)
+                .as("Last product of section should be opened")
+                .isEqualTo(productPage.getProductName());
+
+        productPage
+                .addProductToBasket()
+                .goToCheckout()
+                .didCheckoutPageOpenCorrectly();
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/CheckoutPageTest.java:30]: create a method with boolean return type for verification, that checkout page opened, it could be checked modal (page) title and some necessary element on it

Comment (Nikodych) [reply src/test/java/com/softserveinc/ita/CheckoutPageTest.java:30]: and verify it with assert

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java (hunk 7 above)
Comment (PKuravskyi) [reply src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java:78]: i will probably ignore this problem because i can't reproduce this error and can't test if this method works or not, so im just wasting my time on this. Probably better to say that test was 'Blocked' if this error shows because it's Rozetka's error.

**Comment by PKuravskyi** [comment]
Check comment to Rozetka's Error https://github.com/Nikodych/ifa-134.ta/pull/34#discussion_r756748492

File: src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java (hunk 7 above)
Comment (Nikodych) [reply src/main/java/com/softserveinc/ita/pageobjects/CheckoutPage.java]: I got it

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java [hunk 14]
```diff
@@ -11,25 +11,28 @@
...
-    protected CheckoutPage checkoutPage = new CheckoutPage();
+    private final HomePage homePage = new HomePage();
 
     @Test
     public void verifyCheckoutPageFunctionalityTest() {
-        homePage.closeAdBanner();
-        var lastProductOfSection = homePage.getLastProductOfSection("Акційні пропозиції");
+        homePage
+                .closeAdBanner()
+                .openLastProductOfSection("Акційні пропозиції");
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/CheckoutPageTest.java:20]: extract into variable: "Акційні пропозиції", testSection or something like this

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java [hunk 15]
```diff
@@ -11,25 +11,28 @@
...
+
+        var lastProductOfSection = homePage.getLastProductTitleOfSection("Акційні пропозиції");
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/CheckoutPageTest.java:22]: how it should work, before this you opening product

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java [hunk 16]
```diff
@@ -11,25 +11,28 @@
...
+        var productPage = new ProductPage();
 
         assertThat(lastProductOfSection)
                 .as("Last product of section should be opened")
-                .isEqualTo(productPage.getProductName());
+                .isEqualTo(productPage.getProductTitle());
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/CheckoutPageTest.java:27]: replace it with value in assertThat

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java [hunk 17]
```diff
@@ -11,25 +11,28 @@
...
 
         productPage
                 .addProductToBasket()
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/CheckoutPageTest.java:30]: have you checked your tests after merging master? I see that this method renamed

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java (hunk 17 above)
Comment (PKuravskyi) [reply src/test/java/com/softserveinc/ita/CheckoutPageTest.java:30]: yeah i did

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java [hunk 18]
```diff
@@ -15,19 +15,22 @@
...
-        homePage
-                .closeAdBanner()
-                .openLastProductOfSection("Акційні пропозиції");
+        var testSectionName = "Акційні пропозиції";
+
+        homePage.closeAdBanner();
+
+        var lastProductOfSection = homePage.getLastProductTitleOfSection(testSectionName);
+
+        homePage.openLastProductOfSection(testSectionName);
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/CheckoutPageTest.java:24]: here you can create var productPage, you will get it from openLastProductOfSection

File: src/test/java/com/softserveinc/ita/CheckoutPageTest.java [hunk 19]
```diff
@@ -15,19 +15,22 @@
...
 
-        var lastProductOfSection = homePage.getLastProductTitleOfSection("Акційні пропозиції");
         var productPage = new ProductPage();
```
Comment (Nikodych) [inline src/test/java/com/softserveinc/ita/CheckoutPageTest.java:26]: and delete this line

File: src/main/java/com/softserveinc/ita/pageobjects/HomePage.java [hunk 20]
```diff
@@ -24,7 +24,7 @@ public ProductPage openLastProductOfSection(String sectionName) {
         var lastProductOfSection = getLastProductTitleOfSection(sectionName);
 
         $x(format(GOODS_SECTION_TEMPLATE, sectionName))
-                .$x(format("./following-sibling::ul//a[contains(@title,'%s')]", lastProductOfSection))
+                .$x(format("./following-sibling::ul//a[contains(@title,\"%s\")]", lastProductOfSection))
```
Comment (Nikodych) [inline src/main/java/com/softserveinc/ita/pageobjects/HomePage.java:27]: why did you change to \"%s\" ?

**Review by Nikodych — CHANGES_REQUESTED** [review]

File: src/main/java/com/softserveinc/ita/pageobjects/HomePage.java (hunk 20 above)
Comment (PKuravskyi) [reply src/main/java/com/softserveinc/ita/pageobjects/HomePage.java:27]: if product's name consists apostrophe then my contains is failing

**Comment by PKuravskyi** [comment]
Added new property in pom.xml for maven, otherwise all tests in maven execute in encoding 'cp1251' which fails tests

File: pom.xml [hunk 21]
```diff
@@ -9,6 +9,7 @@
     <version>1.0</version>
 
     <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
```
Comment (Nikodych) [inline pom.xml:12]: nice catch

**Review by Nikodych — APPROVED** [review]
